### PR TITLE
[FEATURE] NPM should use save-exact flags

### PR DESCRIPTION
**Pull request by chadhietala**
NPM chooses not to do this by default. I think is a mistake because of issues with fuzzy matching dependencies.  We have an opportunity to lock dependencies down to real versions in Ember CLI by default.

### CODE REVIEW

**Comment by rwjblue** [comment]
Can you describe what this flag is for, and why we should use it?

**Comment by chadhietala** [comment]
> --save-exact: Saved dependencies will be configured with an exact version rather than using npm's default semver range operator.

So instead of npm flavored semver where they add range operators, you will get a definitive version. For example if you do `npm i glob --save --save-exact` it will download `4.3.2` to node_modules and place `glob: "4.3.2"` into the package.json. It is now guaranteed that your application will get version `4.3.2`.

When you just run `npm i glob --save` you open your self up to minor releases because what goes into your package.json is `glob: ^4.3.2`. This has the potential of having unexpected breaking changes in your CI box or staging environment.

Basically having dependencies that mutate is probably a bad idea and is harder to reason about then just a install version that is immutable.

**Comment by stefanpenner** [comment]
since npm lacks a usable "lock" file (shrinkwrap is deficient) I am for this.

@rwjblue thoughts?

**Comment by rwjblue** [comment]
My concern is that bugfixe releases from a semver bound package (> 1.0.0) would then require app developers to manually research and update. 

Using exact versions for < 1.0.0 packages absolutely makes sense, but for > 1.0.0 I personally would prefer ^1.0.0. 

I'm happy to give this change a chance though, if yall feel strongly about it.

**Comment by chadhietala** [comment]
I think there is the possibility of doing a periodic top level `npm outdated` check during either `ember serve` or when the developer does `ember init` or `ember docter` (TBD).

The Ember ecosystem does an awesome job at semver and have a very structured and no surprises way of bumping versions. That guarantee is not a constant outside of the Ember community though. I think Ember CLI can be the abstraction to mitigate some of the NPM craziness by introduce more sane defaults.

__strong tag for strong opinion__ :smiley:

**Comment by rwjblue** [comment]
Then lets give it a go.
